File: genai_service.py
import os
import requests
from cerebras.cloud.sdk import Cerebras
from dotenv import load_dotenv

load_dotenv()

def get_prompt_template(request_type, data):
    ingredients = data.get("ingredients", "")
    allergies = data.get("allergies", "")
    budget = data.get("budget", "")
    leftovers = data.get("leftovers", "")
    preferences = data.get("preferences", "")
    if request_type == "recipe":
        return f"Generate a recipe with ingredients: {ingredients}. Preferences: {preferences} Allergies: {allergies}. Provide JSON with recipe_name, ingredients, instructions."
    elif request_type == "mealPlan":
        return f"Generate a 7-day meal plan considering: {ingredients}, Preferences: {preferences}, Allergies: {allergies}, Budget: {budget}. Return JSON with name, mealPlan, shoppingList."
    elif request_type == "leftoverOptimize":
        return f"Suggest recipes using these leftovers: {leftovers}."
    # Extend as needed
    return "Invalid prompt."

class GenAIService:
    def __init__(self):
        self.cerebras_api_key = os.getenv("CEREBRAS_API_KEY")
        self.groq_api_key = os.getenv("GROQ_API_KEY")

    def has_valid_api_key(self):
        return bool(self.cerebras_api_key or self.groq_api_key)

    def format_prompt(self, request_type, data):
        return get_prompt_template(request_type, data)

    def generate_mock_response(self, prompt):
        return '{"recipe_name": "Mock Recipe", "ingredients": ["ingredient1","ingredient2"], "instructions": ["Step 1: Do something","Step 2: Done."], "metadata": {"prepTime":"15m", "cookTime":"25m"}}'

    def generate_recipe(self, prompt, options=None):
        if not self.has_valid_api_key():
            print("No valid API key, using mock response")
            return self.generate_mock_response(prompt)
        # Try Cerebras first
        if self.cerebras_api_key:
            try:
                return self.call_cerebras_api(prompt, options)
            except Exception as e:
                print(f"Cerebras API Error: {str(e)}. Trying Groq provider...")
        # Try Groq as fallback
        if self.groq_api_key:
            try:
                return self.call_groq_api(prompt, options)
            except Exception as e:
                print(f"Groq API Error: {str(e)}. Using mock.")
        # If both fail, fallback to mock
        return self.generate_mock_response(prompt)

    def call_cerebras_api(self, prompt, options=None):
        try:
            from cerebras.cloud.sdk import Cerebras
        except ImportError:
            raise ImportError("You must install cerebras-cloud SDK using `pip install cerebras-cloud`")
        model_name = "llama-4-scout-17b-16e-instruct"
        client = Cerebras(api_key=self.cerebras_api_key)
        system_message = "You are a professional chef and nutritionist AI assistant. Provide detailed, accurate, and helpful recipe suggestions with clear instructions and time required at each step, ingredient lists, and nutritional information."
        messages = [
            {"role": "system", "content": system_message},
            {"role": "user", "content": prompt}
        ]
        chat_completion = client.chat.completions.create(
            messages=messages,
            model=model_name,
            max_tokens=options.get("max_tokens", 2000) if options else 2000,
            temperature=options.get("temperature", 0.7) if options else 0.7,
        )
        return chat_completion.choices[0].message.content

    def call_groq_api(self, prompt, options=None):
        try:
            from groq import Groq
        except ImportError:
            raise ImportError("You must install groq SDK using `pip install groq`")
        client = Groq(api_key=self.groq_api_key)
        messages = [
            {
                "role": "system", 
                "content": "You are a professional chef and nutritionist AI assistant. Provide detailed, accurate, and helpful recipe suggestions with clear instructions and time required at each step, ingredient lists, and nutritional information."
            },
            {
                "role": "user",
                "content": prompt
            }
        ]
        completion = client.chat.completions.create(
            model="llama-3.1-70b-versatile",
            messages=messages,
            temperature=options.get("temperature", 0.7) if options else 0.7,
            max_completion_tokens=options.get("max_tokens", 2000) if options else 2000,
            top_p=1,
            reasoning_effort="medium",
            stream=False,
            stop=None
        )
        return completion.choices[0].message.content

File: services/genaiService.js
const axios = require('axios');

class GenAIService {
  constructor() {
    this.provider = process.env.GENAI_PROVIDER || 'cerebras';
    this.cerebrasApiKey = process.env.CEREBRAS_API_KEY;
    this.llamaApiKey = process.env.LLAMA_API_KEY;
    this.openaiApiKey = process.env.OPENAI_API_KEY;
  }

  async generateRecipe(prompt, options = {}) {
    try {
      let response;
      
      switch (this.provider) {
        case 'cerebras':
          response = await this.callCerebrasAPI(prompt, options);
          break;
        case 'llama':
          response = await this.callLlamaAPI(prompt, options);
          break;
        case 'openai':
          response = await this.callOpenAIAPI(prompt, options);
          break;
        default:
          throw new Error(`Unknown GenAI provider: ${this.provider}`);
      }

      return response;
    } catch (error) {
      console.error('GenAI API Error:', error.message);
      throw new Error(`Failed to generate content: ${error.message}`);
    }
  }

  async callCerebrasAPI(prompt, options) {
    // Cerebras API integration
    const endpoint = 'https://api.cerebras.ai/v1/chat/completions';
    
    const requestData = {
      model: 'llama3.1-8b',
      messages: [
        {
          role: 'system',
          content: 'You are a professional chef and nutritionist AI assistant. Provide detailed, accurate, and helpful recipe suggestions with clear instructions, ingredient lists, and nutritional information.'
        },
        {
          role: 'user',
          content: prompt
        }
      ],
      max_tokens: options.maxTokens || 2000,
      temperature: options.temperature || 0.7,
      stream: false
    };

    try {
      const response = await axios.post(endpoint, requestData, {
        headers: {
          'Authorization': `Bearer ${this.cerebrasApiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 30000
      });

      return response.data.choices[0].message.content;
    } catch (error) {
      if (error.response?.status === 401) {
        throw new Error('Invalid Cerebras API key. Please check your configuration.');
      }
      throw new Error(`Cerebras API error: ${error.message}`);
    }
  }

  async callLlamaAPI(prompt, options) {
    // Meta LLaMA API integration (placeholder - adjust based on actual API)
    const endpoint = 'https://api.llama-api.com/chat/completions';
    
    const requestData = {
      model: 'llama-2-70b-chat',
      messages: [
        {
          role: 'system',
          content: 'You are a professional chef and nutritionist AI assistant. Provide detailed, accurate, and helpful recipe suggestions.'
        },
        {
          role: 'user',
          content: prompt
        }
      ],
      max_tokens: options.maxTokens || 2000,
      temperature: options.temperature || 0.7
    };

    try {
      const response = await axios.post(endpoint, requestData, {
        headers: {
          'Authorization': `Bearer ${this.llamaApiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 30000
      });

      return response.data.choices[0].message.content;
    } catch (error) {
      throw new Error(`LLaMA API error: ${error.message}`);
    }
  }

  async callOpenAIAPI(prompt, options) {
    // OpenAI API as fallback
    const endpoint = 'https://api.openai.com/v1/chat/completions';
    
    const requestData = {
      model: 'gpt-3.5-turbo',
      messages: [
        {
          role: 'system',
          content: 'You are a professional chef and nutritionist AI assistant. Provide detailed, accurate, and helpful recipe suggestions.'
        },
        {
          role: 'user',
          content: prompt
        }
      ],
      max_tokens: options.maxTokens || 2000,
      temperature: options.temperature || 0.7
    };

    try {
      const response = await axios.post(endpoint, requestData, {
        headers: {
          'Authorization': `Bearer ${this.openaiApiKey}`,
          'Content-Type': 'application/json'
        },
        timeout: 30000
      });

      return response.data.choices[0].message.content;
    } catch (error) {
      throw new Error(`OpenAI API error: ${error.message}`);
    }
  }

  // Helper method to format prompts for different use cases
  formatPrompt(type, data) {
    const prompts = {
      recipe: `Generate a detailed recipe using these ingredients: ${data.ingredients}. 
               Consider dietary restrictions: ${data.allergies || 'none'}. 
               Budget constraint: ${data.budget || 'no limit'}. 
               Available leftovers: ${data.leftovers || 'none'}.
               
               Please provide:
               1. Recipe name and description
               2. Complete ingredient list with quantities
               3. Step-by-step instructions
               4. Estimated preparation and cooking time
               5. Servings
               6. Nutritional information (approximate)
               7. Cost estimate if budget was provided
               
               Format the response in JSON with clear sections.`,
      
      mealPlan: `Create a 7-day meal plan with these constraints:
                 Ingredients available: ${data.ingredients}
                 Dietary restrictions: ${data.allergies || 'none'}
                 Budget: ${data.budget || 'flexible'}
                 Leftovers to use: ${data.leftovers || 'none'}
                 
                 Provide a day-by-day meal plan with breakfast, lunch, and dinner suggestions.
                 Include a consolidated shopping list.
                 Format as JSON.`,
      
      allergyCheck: `Analyze this recipe for potential allergens and provide substitutions:
                     Recipe: ${data.recipe}
                     Known allergies: ${data.allergies}
                     
                     Highlight any allergens found and suggest safe alternatives.
                     Format as JSON with warnings and substitutions.`,
      
      budgetOptimize: `Optimize this recipe for budget constraints:
                       Recipe: ${data.recipe}
                       Budget limit: ${data.budget}
                       
                       Suggest cost-effective alternatives and provide estimated total cost.
                       Format as JSON.`,
      
      leftoverOptimize: `Create new recipes using these leftovers:
                         Available leftovers: ${data.leftovers}
                         Additional ingredients available: ${data.ingredients || 'basic pantry items'}
                         
                         Suggest creative recipes that minimize waste.
                         Format as JSON.`
    };

    return prompts[type] || prompts.recipe;
  }
}

module.exports = new GenAIService();